### ISSUE TITLE
L162.defaultYieldRate:  left_join_no_match

### ISSUE THREAD

**Issue by martikj2** [JGCRI/gcamdata#1138]
Tried to build the data system on my Ubuntu 16.04 machine. As I run driver() in Rstudio (as explained in the documentation https://github.com/JGCRI/gcamdata/wiki/Getting-Started), everything seemed to run smoothly until there was a crash:

"module_aglu_LB162.ag_prodchange_R_C_Y_GLU_irr"
Error in left_join_error_no_match(., L162.defaultYieldRate, by = c("year",  : 
  left_join_no_match: NA values in new data columns
Called from: left_join_error_no_match(., L162.defaultYieldRate, by = c("year", 
    "GCAM_commodity"))

I made no modification to the code not the data provided. Any idea how to fix this?

**Comment by bpbond** [comment]
Hi @martikj2, my first question is what versions are you running? See #1125 for example - are you running the default version included with GCAM?

**Comment by pkyle** [comment]
This is strange that it seems to be dependent on operating systems and/or library versions; the reason that the join is failing is that the LHS has "year" as a double while the RHS has it as an integer. The first instance that I can find in that code file where a year column is a double is in line 216:
`repeat_add_columns(tibble::tibble(year = aglu.SPEC_AG_PROD_YEARS)) %>%`
So, it would likely help out if you just specified that as an integer:
`repeat_add_columns(tibble::tibble(year = as.integer(aglu.SPEC_AG_PROD_YEARS))) %>%`
But my guess is, if all of the gcamdata development took place in environments where doubles and ints could be joined seamlessly, there will likely be many other instances of this. Might want to figure out if a recent change to dplyr modified the behavior here.

**Comment by martikj2** [comment]
I tried the gcamdata provided in the gcam-core as well as the one from https://github.com/JGCRI/gcamdata

**Comment by pkyle** [comment]
That's not surprising that they would have the same outcome as the two repositories have the same code. What version of dplyr are you running? From the DESCRIPTION we require `dplyr (>= 0.7.0)`, I run 0.7.8, and the [current one](https://github.com/tidyverse/dplyr/releases) is 0.8.3.
It may also be helpful to try this test code in your R console, just to see what happens if you `left_join` two tables by a variable that's an integer in one and a double in the other. (You'll need to load dplyr and tibble if you haven't already)
```
tmp <- tibble(year = as.integer(2001:2002), value = 1:2)
tmp1 <- tibble(year = as.double(2001:2002), id = c("red", "blue"))
left_join(tmp, tmp1, by = "year")
```
If that produces NA values in the "id" column, then that identifies the problem for sure. And, if you're running the most recent version of dplyr, the easiest fix for now may be to instead use an older one. It's hard for us to keep up with all of the changes they make to the tidyverse packages!

**Comment by martikj2** [comment]
Thanks for your help, but sadly I am still struggling with this. The left_join seems to work fine...I get
# A tibble: 2 x 3
   year value id   
  <dbl> <int> <chr>
1  2001     1 red  
2  2002     2 blue 

I tried downgrading dplyr to 0.7.8. However, then I ran into installation problem already in gcamdata since it wanted the newer version of dplyr in any case. Maybe lots of other things in the environment need to be downgraded as well, not sure which and to which versions. (My rstudio install is very recent.)

If I am not mistaken, to run gcam I would basically just need those xml files that gcamdata generates. Such sets are provided in windows download. Is it maybe possible to provide functioning sets in gcam-core as well and in this way bypass  the problem in building the data system on my own?

**Comment by bpbond** [comment]
Can you provide us with a `sessionInfo()` output please, from right after the error occurs?

**Comment by martikj2** [comment]
sessionInfo() after driver() crash looks like this....

R version 3.2.3 (2015-12-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.6 LTS
locale:
 [1] LC_CTYPE=en_US.UTF-8      LC_NUMERIC=C              LC_TIME=en_DK.utf8       
 [4] LC_COLLATE=en_US.UTF-8    LC_MONETARY=fi_FI.utf8    LC_MESSAGES=en_US.UTF-8  
 [7] LC_PAPER=fi_FI.utf8       LC_NAME=C                 LC_ADDRESS=C             
[10] LC_TELEPHONE=C            LC_MEASUREMENT=fi_FI.utf8 LC_IDENTIFICATION=C      
attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] gcamdata_5.1
loaded via a namespace (and not attached):
 [1] Rcpp_1.0.2        tidyr_1.0.0       zeallot_0.1.0     crayon_1.3.4      dplyr_0.8.3      
 [6] assertthat_0.2.1  R6_2.4.0          lifecycle_0.1.0   backports_1.1.4   magrittr_1.5     
[11] pillar_1.4.2      stringi_1.4.3     rlang_0.4.0       data.table_1.12.2 rstudioapi_0.10  
[16] ellipsis_0.3.0    vctrs_0.2.0       tools_3.2.3       readr_1.3.1       glue_1.3.1       
[21] purrr_0.3.2       hms_0.5.1         pkgconfig_2.0.3   tidyselect_0.2.5  tibble_2.1.3

**Comment by pralitp** [comment]
This has to do with `tidyr 1.0.0` changing behavior from version `0.8.x`.  Note this particular one causes a crash but there are several other places where it doesn't crash but continues on with slightly different values.

I have fixes which hopefully will get merged (and should be backward compatible with older versions of `tidyr`) soon but right now it is tied up with some other work.  You can take a look at c09df9afb18724b7fc96313b7f472f0110107bdd and d830444a9b42891187e02fae15fbb25581a14924 if you need to get this working more urgently (note some of these R files may not be in your branch or master in which case just ignore them).

**Comment by BizYoder** [comment]
@martikj2 Did you figure out how to solve this? I'm getting the same error.

**Comment by pkyle** [comment]
I'm pretty sure the following will work, as the two commits are consecutive and shouldn't conflict with any other changes you may have made:
```
git cherry-pick c09df9a..d830444
```

**Comment by pralitp** [comment]
This should have closed with #1140
